### PR TITLE
Add sdxl prompt embeddings

### LINKED ISSUE #3970: SDXL prompt embeddings not working
### Describe the bug

```
pipe = StableDiffusionXLPipeline.from_pretrained(
                    model_id, torch_dtype=torch.float16, use_safetensors=True, variant="fp16", use_auth_token=hf_token).to("cuda")
cond, uncond = text_embeddings(
    self.pipe, args_dict["prompt"], args_dict["negative_prompt"], clip_skip=2)
del args_dict["prompt"]
del args_dict["negative_prompt"]
args_dict["prompt_embeds"] = cond
args_dict["negative_prompt_embeds"] = uncond
pipe(**args_dict)
```

Gives error
```

File "/opt/conda/envs/ldm/lib/python3.11/site-packages/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py", line 402, in encode_prompt
pooled_prompt_embeds = pooled_prompt_embeds.repeat(1, num_images_per_prompt).view(
                           ^^^^^^^^^^^^^^^^^^^^
 UnboundLocalError: cannot access local variable 'pooled_prompt_embeds' where it is not associated with a value
```

### Reproduction

SDXL

### Logs

_No response_

### System Info

pip3 install git+https://github.com/huggingface/diffusers.git

### Who can help?

@patrickvonplaten

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Make sure prompt embeddings can be passed to SD-XL.

Fixes #3970

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion_xl/__init__.py [hunk 1]
```diff
@@ -17,13 +16,9 @@ class StableDiffusionXLPipelineOutput(BaseOutput):
         images (`List[PIL.Image.Image]` or `np.ndarray`)
             List of denoised PIL images of length `batch_size` or numpy array of shape `(batch_size, height, width,
             num_channels)`. PIL images or numpy array present the denoised images of the diffusion pipeline.
-        nsfw_content_detected (`List[bool]`)
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion_xl/__init__.py:20]: Why are removing this?

**Review by sayakpaul — COMMENTED** [review]
I would have expected to see a change in `encode_prompt()` to accommodate the pooled embeddings. But didn't notice it.

What am I missing out on? We should maybe also include tests for those.

**Review by pcuenca — COMMENTED** [review]
I don't fully follow. How would this be used with the `compel` library?

**Comment by patrickvonplaten** [comment]
Having chatted offilne with @pcuenca this is the only way really we can do things. 

If the user passes `prompt_embeds` we assume that the text encoder is not needed anymore and that the user does **not** have to pass `prompt`. This however means that the user has to pass `pooled_prompt_embeds`.

**Comment by patrickvonplaten** [comment]
Merging

**Comment by adhikjoshi** [comment]
So, pooled_prompt_embeds is also needed with prompt_embeds

We have understanding of creating embeds, but pooled embeds will be same or they are different?

Can we pass same prompt embeddings in pooled and prompt_embeds?

If not, how can we create pooled embeds? Can compel work?

**Comment by sayakpaul** [comment]
Hi @adhikjoshi. 

If you check out 

https://github.com/huggingface/diffusers/blob/78922ed7c7e66c20aa95159c7b7a6057ba7d590d/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L242

you will notice the code for computing the pooled prompt embeddings. 

Let me know if that helps.

**Comment by bghira** [comment]
```python
                # We are only ALWAYS interested in the pooled output of the final text encoder
                negative_pooled_prompt_embeds = negative_prompt_embeds[0]

```
seems like it's not needed and we can just pass in the negative prompt embeds etc?

**Comment by bghira** [comment]
docs for encode_prompt say prompt is optional, but signature has no default value of None

**Pull request merged**